### PR TITLE
Constraints Support 2

### PR DESCRIPTION
**Pull request by vittvolt**
## Overview
The code adds some basic constraint support to the Peloton database. Our main changes include adding NOT NULL, DEFAULT, CHECK, consolidating PRIMARY KEY and UNIQUE constraints and adding appropriate test cases.

## Changes
The primary changes we have made include:

~/src/catalog : The column, constraint and schema definitions have been changed to accommodate for checking the constraints on each tuple.
~/src/storage : The data_table.cpp file has been modified to add the new constraint checking routines when allocating a new slot.
~/test/catalog : Added new test cases for constraint tests. Note that constraints_tests_utils.cpp should be ignored, refer to constraints_tests_utils.h for implementation.
~/src/planner : some files have been changed to support exception handling and default case improvement: plan_executor.cpp and the other executors
~/src/parser: create_statement.h
~/src/tcop: tcop.cpp are modified for exception handling as well.

## Issues
- Foreign key support with cascade on delete / update is not ready.

- The stats_test may call InsertTuple directly without exception handling, as a result the stats_test could not pass.

- Parser support is mostly ready.

- Some more improvement and testing needed.

- new constraint data structure needs to be added in create_statement.h for the parser to set appropriate constraints

### CODE REVIEW

File: test/catalog/constraints_test.cpp [hunk 1]
```diff
@@ -43,8 +65,6 @@ namespace test {
 class ConstraintsTests : public PelotonTest {};
 
 // FIXME: see the explanation rpc_client_test and rpc_server_test
-TEST_F(ConstraintsTests, BlankTest) {}
-/*
 #ifdef NOTNULL_TEST
```
Comment (vittvolt) [inline test/catalog/constraints_test.cpp]: It's mostly for debugging purpose. Sometimes we just want to run a few tests to check for memory / pointer issues.

File: .gitignore [hunk 2]
```diff
@@ -129,6 +129,11 @@ doc/
 *.iml
 
 #==============================================================================#
+# VS Code files
+#==============================================================================#
+.vscode
+
```
Comment (HrWangChengdu) [inline .gitignore]: Is this change necessary? Why it would generate .vscode file?

Comment (vittvolt) [inline .gitignore]: @HrWangChengdu I'm using vs code editor. So...

File: test/executor/testing_executor_util.cpp [hunk 3]
```diff
@@ -120,6 +120,22 @@ catalog::Column TestingExecutorUtil::GetColumnInfo(int index) {
...
     } break;
 
+    case 4: {
+      auto column = catalog::Column(
+          type::Type::INTEGER, type::Type::GetTypeSize(type::Type::INTEGER),
+          "COL_E", is_inlined);
+      column.AddConstraint(catalog::Constraint(ConstraintType::DEFAULT,
+                                               not_null_constraint_name));
+    } break;
+
```
Comment (HrWangChengdu) [inline test/executor/testing_executor_util.cpp]: Could you add some comments for the magic number?

File: test/catalog/constraints_tests_util.cpp [hunk 4]
```diff
@@ -102,6 +102,17 @@ catalog::Column ConstraintsTestsUtil::GetColumnInfo(int index) {
       return column;
     } break;
 
+    case 114: {
+      auto column = catalog::Column(type::Type::INTEGER,
+                                    25,  // Column length.
```
Comment (HrWangChengdu) [inline test/catalog/constraints_tests_util.cpp]: Could you add more comments for "25"?  "Column length"  doesn't really help me understand the code.

File: .gitignore (hunk 2 above)
Comment (HrWangChengdu) [inline .gitignore]: @vittvolt  I guess a better way is to keep the changes to .gitignore local

File: src/include/catalog/schema.h [hunk 5]
```diff
@@ -171,6 +173,28 @@ class Schema : public Printable {
...
+      if (constraint.GetType() == ConstraintType::DEFAULT) {
+        return constraint.getDefaultValue();
+      }
+    }
+
+    return nullptr;
+  }
+
   // Add constraint for column by id
   inline void AddConstraint(oid_t column_id,
```
Comment (NasrinJaleel93) [inline src/include/catalog/schema.h]: Dead Code?

File: src/tcop/tcop.cpp [hunk 6]
```diff
@@ -234,8 +234,18 @@ executor::ExecuteResult TrafficCop::ExecuteStatementPlan(
   // skip if already aborted
   if (curr_state.second != ResultType::ABORTED) {
     PL_ASSERT(txn);
-    p_status = executor::PlanExecutor::ExecutePlan(plan, txn, params, result,
+    // Handle constraint exceptions
+    try {
+      p_status = executor::PlanExecutor::ExecutePlan(plan, txn, params, result,
                                                  result_format);
+    } catch (Exception &e) {
+      auto &txn_manager = concurrency::TransactionManagerFactory::GetInstance();
```
Comment (NasrinJaleel93) [inline src/tcop/tcop.cpp]: I am not sure if you need a new transaction object here?

**Comment by apavlo** [comment]
This PR doesn't build. Have any of the comments from the second code review been addressed?

Matt (@opensky2) should fix his commit email address so that it links to his account.

File: src/executor/update_executor.cpp [hunk 7]
```diff
@@ -290,6 +295,14 @@ bool UpdateExecutor::DExecute() {
             }
             transaction_manager.SetTransactionResult(current_txn,
                                                      ResultType::FAILURE);
+            if (fk_failure) {
+              throw ConstraintException("FOREIGN KEY constraint violated : " +
+                               std::string(new_tuple.GetInfo()));
+            } else {
+              throw ConstraintException("Secondary index constraint violated : " +
```
Comment (opensky2) [inline src/executor/update_executor.cpp]: Yes, these ConstraintExceptions are caught in the traffic cop, which then updates the transaction result appropriately.

File: src/executor/plan_executor.cpp [hunk 8]
```diff
@@ -82,7 +82,19 @@ ExecuteResult PlanExecutor::ExecutePlan(const planner::AbstractPlan *plan,
...
+        // The best way to handle constraint violations is NOT using
+        // exceptions
+        try {
+          status = executor_tree->Execute();
+        } catch (Exception &e) {
+          p_status.m_processed = executor_context->num_processed;
+          p_status.m_result = ResultType::FAILURE;
+          p_status.m_result_slots = nullptr;
+          CleanExecutorTree(executor_tree.get());
+          throw e;
```
Comment (opensky2) [inline src/executor/plan_executor.cpp]: Yes, these ConstraintExceptions are caught in the traffic cop, which then updates the transaction result appropriately.

File: src/include/catalog/foreign_key.h [hunk 9]
```diff
@@ -29,50 +29,48 @@ namespace catalog {
 // Stores info about foreign key constraints, like the sink table id etc.
 class ForeignKey {
  public:
```
Comment (opensky2) [inline src/include/catalog/foreign_key.h]: The reason we have strings for table and column names here is that the FK information is added to the CreatePlan for a table in the planner, at which point aren't passed anything but the strings that came in from the parser.

File: src/include/catalog/schema.h (hunk 5 above)
Comment (vittvolt) [inline src/include/catalog/schema.h]: Two AddConstraint functions, one for a given column id, and another to deal with column names.

File: src/tcop/tcop.cpp (hunk 6 above)
Comment (vittvolt) [inline src/tcop/tcop.cpp]: I think it's needed to introduce minimum change to the original code.

**Comment by vittvolt** [comment]
@apavlo As I posted in the stats team's PR, the constraint exception is not handled in stats_test, therefore cannot pass the test.

**Comment by apavlo** [comment]
@vittvolt Can you fix your commits so that they are properly linked to your account? I want to make sure that everyone gets credit for their contributions.

https://github.com/cmu-db/peloton/pull/660/commits/84a0213e38a35b24444b7a26e99c442b79d2bcf0

**Comment by vittvolt** [comment]
@apavlo I ran a few scripts for setting the authors and I believe all the commits have been properly addressed.

**Comment by apavlo** [comment]
Looks better. Thanks!

**Comment by apavlo** [comment]
I am going to have to merge this code separately in another branch. We need to remove the .docx file from the history. I also have fixes for the failing test cases.

**Comment by apavlo** [comment]
Merged in #794
